src/dkg.rs

```diff
@@ -0,0 +1,512 @@
+    EvaluationDomain, Polynomial,
+};
+use ed25519_dalek as ed25519;
+
+// DKG parameters
+#[derive(Copy, Clone, Debug)]
+pub struct Params {
+    pub failure_threshold: u32,  // failure threshold
+    pub security_threshold: u32, // threshold
+    pub total_weight: u32,       // total weight
+}
+
+#[derive(Debug)]
+pub struct ParticipantBuilder {
+    pub ed_key: ed25519::PublicKey,
+    pub dh_key: dh::AsymmetricPublicKey,
+    pub stake: u64,
+}
+
+#[derive(Clone, Debug)]
+pub struct Participant {
+    pub ed_key: ed25519::PublicKey,
+    pub dh_key: dh::AsymmetricPublicKey,
+    pub weight: u32,
+    pub share_range: std::ops::Range<usize>,
+    pub share_domain: fastpoly::SubproductDomain, // subproduct tree of polynomial with zeros at share_domain
+    pub domain_commitment: Option<G2Affine>, //Commitment to subproduct domain}
+}
+#[derive(Debug)]
+pub enum State {
+    Init {
+        participants: Vec<ParticipantBuilder>,
+    },
+    Sharing {
+        finalized_weight: u32,
+    },
+    Success,
+    Failure,
+}
+
+pub struct Context {
+    pub tau: u32,
+    pub dh_key: dh::AsymmetricKeypair,
+    pub ed_key: ed25519::Keypair,
+    pub params: Params,
+    pub participants: Vec<Participant>,
+    pub vss: BTreeMap<u32, syncvss::sh::Context>,
+    pub domain: ark_poly::Radix2EvaluationDomain<Scalar>,
+    pub state: State,
+    pub me: usize,
+    pub powers_of_g: Rc<Vec<G1Affine>>,
+    pub powers_of_h: Rc<Vec<G2Affine>>,
+    pub beta_h: G2Affine,
+}
+
+impl Context {
+    pub fn new<R: rand::Rng + rand::CryptoRng + Sized>(
+        tau: u32,
+        ed_key: ed25519::Keypair,
+        params: &Params,
+        powers_of_g: Rc<Vec<G1Affine>>,
+        powers_of_h: Rc<Vec<G2Affine>>,
+        beta_h: G2Affine,
+        rng: &mut R,
+    ) -> Result<Context, anyhow::Error> {
+        let domain = ark_poly::Radix2EvaluationDomain::<Scalar>::new(
+            params.total_weight as usize,
+        )
+        .ok_or_else(|| anyhow::anyhow!("unable to construct domain"))?;
+
+        Ok(Context {
+            tau,
+            dh_key: dh::AsymmetricKeypair::new(rng),
+            ed_key,
+            params: *params,
+            participants: vec![],
+            vss: BTreeMap::new(),
+            domain,
+            state: State::Init {
+                participants: vec![],
+            },
+            me: 0, // TODO: invalid value
+            powers_of_g,
+            powers_of_h,
+            beta_h,
+        })
+    }
+    pub fn final_key(&self) -> G1Affine {
+        use crate::syncvss::sh::State as VSSState;
+        use ark_ff::Zero;
+        self.vss
+            .iter()
+            .filter_map(|(_, vss)| {
+                if let VSSState::Success { final_secret } = vss.state {
+                    Some(final_secret)
+                } else {
+                    None
+                }
+            })
+            // .sum() //TODO: it would be nice to use Sum trait here, but not implemented for G1Affine
```

Comment:
created an arkworks issue, might grab it
https://github.com/arkworks-rs/algebra/issues/288